File: src/http/admin/lastfm.go
package admin

import (
	"encoding/json"
	"net/http"
	"tapesonic/lastfm"
	"tapesonic/model"
	"time"
)

type LastFmAuthLinkRs struct {
	Url   string
	Token string
}

func PostLastFmAuthLink(auth *authenticator, lastfm *lastfm.LastFmService) WebappHandler {
	return func(r *http.Request) (any, error) {
		_, err := auth.Authenticate(r)
		if err != nil {
			return nil, err
		}

		link, err := lastfm.CreateAuthLink()
		if err != nil {
			return nil, err
		}

		return LastFmAuthLinkRs{
			Url:   link.Url,
			Token: link.Token,
		}, nil
	}
}

type LastFmSessionRs struct {
	Username            string
	IsScrobblingEnabled bool
	UpdatedAt           time.Time
}

func toLastFmSessionRs(session lastfm.LastFmSession) LastFmSessionRs {
	return LastFmSessionRs{
		Username:            session.Username,
		IsScrobblingEnabled: session.IsScrobblingEnabled,
		UpdatedAt:           session.UpdatedAt.Unwrap(),
	}
}

func GetLastFmSession(auth *authenticator, lastfm *lastfm.LastFmService) WebappHandler {
	return func(r *http.Request) (any, error) {
		user, err := auth.Authenticate(r)
		if err != nil {
			return nil, err
		}

		session, err := lastfm.GetCurrentSession(user)
		if err != nil {
			return nil, err
		}

		if session == nil {
			return nil, model.ErrNotFound
		}

		return toLastFmSessionRs(*session), nil
	}
}

type CreateLastFmSessionRq struct {
	Token string
}

func PostLastFmSession(auth *authenticator, lastfm *lastfm.LastFmService) WebappHandler {
	return func(r *http.Request) (any, error) {
		user, err := auth.Authenticate(r)
		if err != nil {
			return nil, err
		}

		request := CreateLastFmSessionRq{}
		err = json.NewDecoder(r.Body).Decode(&request)
		if err != nil {
			return nil, err
		}

		session, err := lastfm.CreateSession(user, request.Token)
		if err != nil {
			return nil, err
		}

		return toLastFmSessionRs(session), nil
	}
}

func DeleteLastFmSession(auth *authenticator, lastfm *lastfm.LastFmService) WebappHandler {
	return func(r *http.Request) (any, error) {
		user, err := auth.Authenticate(r)
		if err != nil {
			return nil, err
		}

		return nil, lastfm.DeleteSession(user)
	}
}

type LastFmSessionSettingsRq struct {
	IsScrobblingEnabled bool
}

func PutLastFmSessionSettings(auth *authenticator, lastfmSvc *lastfm.LastFmService) WebappHandler {
	return func(r *http.Request) (any, error) {
		user, err := auth.Authenticate(r)
		if err != nil {
			return nil, err
		}

		var rq LastFmSessionSettingsRq
		if err := json.NewDecoder(r.Body).Decode(&rq); err != nil {
			return nil, err
		}

		session, err := lastfmSvc.UpdateSessionSettings(
			user,
			lastfm.SessionSettings{IsScrobblingEnabled: rq.IsScrobblingEnabled},
		)
		if err != nil {
			return nil, err
		}

		return toLastFmSessionRs(session), nil
	}
}

File: webapp/src/api/lastfm.ts
import api from '@/api/api';

export interface LastFmSessionRs {
    Username: string;
    IsScrobblingEnabled: boolean;
    UpdatedAt: string;
}

export interface LastFmAuthLinkRs {
    Url: string;
    Token: string;
}

export interface CreateLastFmSessionRq {
    Token: string;
}

export interface LastFmSessionSettingsRq {
    IsScrobblingEnabled: boolean;
}

export default {
    async getCurrentLastFmSession(): Promise<LastFmSessionRs | null> {
        const response = await fetch(`/api/lastfm/auth`, { method: "GET" });
        return await api.handleNullableResponse(response);
    },
    async createLastFmAuthLink(): Promise<LastFmAuthLinkRs> {
        const response = await fetch(`/api/lastfm/create-auth-link`, { method: "POST" });
        return await api.handleRequiredResponse(response);
    },
    async createLastFmSession(token: string): Promise<LastFmSessionRs> {
        const request: CreateLastFmSessionRq = {
            Token: token,
        };
        const response = await fetch(`/api/lastfm/auth`, { method: "POST", body: JSON.stringify(request) });
        return await api.handleRequiredResponse(response);
    },
    async deleteLastFmSession(): Promise<void> {
        const response = await fetch(`/api/lastfm/auth`, { method: "DELETE" });
        await api.handleRequiredResponse(response);
    },
    async updateLastFmSessionSettings(rq: LastFmSessionSettingsRq): Promise<LastFmSessionRs> {
        const response = await fetch(`/api/lastfm/settings`, { method: "PUT", body: JSON.stringify(rq) });
        return await api.handleRequiredResponse(response);
    },
}
